README.md
# dash-webcam

Dash UI component suite

## Dash

Go to this link to learn about [Dash][].

## Getting started

```sh
# Install dependencies
$ npm install

# Watch source for changes and build to `lib/`
$ npm start
```

## Development

### Demo server

You can start up a demo development server to see a demo of the rendered
components:

```sh
$ builder run demo
$ open http://localhost:9000
```

You have to maintain the list of components in `demo/Demo.react.js`.

### Code quality and tests

#### To run lint and unit tests:

```sh
$ npm test
```

#### To run unit tests and watch for changes:

```sh
$ npm run test-watch
```

#### To debug unit tests in a browser (Chrome):

```sh
$ npm run test-debug
```

1. Wait until Chrome launches.
2. Click the "DEBUG" button in the top right corner.
3. Open up Chrome Devtools (`Cmd+opt+i`).
4. Click the "Sources" tab.
5. Find source files
  - Navigate to `webpack:// -> . -> spec/components` to find your test source files.
  - Navigate to `webpack:// -> [your/repo/path]] -> dash-webcam -> src` to find your component source files.
6. Now you can set breakpoints and reload the page to hit them.
7. The test output is available in the "Console" tab, or in any tab by pressing "Esc".

#### To run a specific test

In your test, append `.only` to a `describe` or `it` statement:

```javascript
describe.only('Foo component', () => {
    // ...
})l
```

### Testing your components in Dash

1. Build development bundle to `lib/` and watch for changes

        # Once this is started, you can just leave it running.
        $ npm start

2. Install module locally (after every change)

        # Generate metadata, and build the JavaScript bundle
        $ npm run install-local

        # Now you're done. For subsequent changes, if you've got `npm start`
        # running in a separate process, it's enough to just do:
        $ python setup.py install

3. Run the dash layout you want to test

        # Import dash-webcam to your layout, then run it:
        $ python my_dash_layout.py


**TODO:** There is a workflow that links your module into `site-packages` which would
make it unnecessary to re-run `2.` on every change: `python setup.py develop`.
Unfortunately, this doesn't seem to work with resources defined in
`package_data`.

See https://github.com/plotly/dash-components-archetype/issues/20


## Installing python package locally

Before publishing to PyPi, you can test installing the module locally:

```sh
# Install in `site-packages` on your machine
$ npm run install-local
```

## Uninstalling python package locally

```sh
$ npm run uninstall-local
```

## Publishing

For now, multiple steps are necessary for publishing to NPM and PyPi,
respectively. **TODO:**
[#5](https://github.com/plotly/dash-components-archetype/issues/5) will roll up
publishing steps into one workflow.

Ask @chriddyp to get NPM / PyPi package publishing accesss.

1. Preparing to publish to NPM

        # Bump the package version
        $ npm version major|minor|patch

        # Push branch and tags to repo
        $ git push --follow-tags

2. Preparing to publish to PyPi

        # Bump the PyPi package to the same version
        $ vi setup.py

        # Commit to github
        $ git add setup.py
        $ git commit -m "Bump pypi package version to vx.x.x"

3. Publish to npm and PyPi

        $ npm run publish-all

## Builder / Archetype

We use [Builder][] to centrally manage build configuration, dependencies, and
scripts.

To see all `builder` scripts available:

```sh
$ builder help
```

See the [dash-components-archetype][] repo for more information.


[Builder]: https://github.com/FormidableLabs/builder
[Dash]: https://github.com/plotly/dash2
[dash-components-archetype]: https://github.com/plotly/dash-components-archetype


CHANGELOG.md
# Change Log for dash-webcam
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased

### Added
- Feature x

### Fixed
- Bug y

## 0.0.1 - 2018-07-23
- Initial release

[Unreleased]: https://github.com/plotly/dash-webcam/v0.0.1...HEAD


usage.py
import dash_webcam
import dash
from dash.dependencies import Input, Output, State
import dash_html_components as html

app = dash.Dash('')

app.scripts.config.serve_locally = True

app.layout = html.Div([
    html.Div(className='row', children=[
        html.P('Original Footage:'),

        dash_webcam.WebcamComponent(
            id='webcam',
            audio=False,
            screenshotFormat="image/jpeg",
            width=500,
            height=300
        )
    ]),

    html.Div(className='row', children=[
        html.P('Returned Footage:'),
        html.Img(id='output')
    ])

])


@app.callback(Output('output', 'src'),
              [Input('webcam', 'screenshot')])
def show_screenshot_image(screenshot):
    return screenshot


if __name__ == '__main__':
    app.run_server(debug=True)


src/components/WebcamComponent.react.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import Webcam from 'react-webcam';

/**
 * ExampleComponent is an example component.
 * It takes a property, `label`, and
 * displays it.
 * It renders an input with the property `value`
 * which is editable by the user.
 */
export default class WebcamComponent extends Component {
    constructor(props) {
        super(props);
        this.state = {value: props.value};

        this.getScreenshot = this.getScreenshot.bind(this);
        this.handleScreenshot = setInterval(this.getScreenshot, 40);  // 25 FPS
    }

    getScreenshot(){
        const {setProps} = this.props;
        const currScreenshot = this.refs.webcam.getScreenshot();

        if (setProps  !== null) {
            setProps({screenshot: currScreenshot});
        }
    }

    updateIntervals(prevProps){
        const {screenshotInterval} = this.props;

        if (this.handleScreenshot === undefined ||
            prevProps.screenshotInterval !== screenshotInterval){
            clearInterval(this.handleScreenshot);
            this.handleScreenshot = setInterval(
                this.getScreenshot,
                screenshotInterval
            );
        }
    }

    componentDidUpdate(prevProps){
        this.updateIntervals(prevProps);
    }

    render() {
        const {
            className,
            audio,
            height,
            width,
            screenshotWidth,
            screenshotFormat,
            screenshotQuality
        } = this.props;

        return (
            <Webcam
                ref="webcam"
                className={className}
                audio={audio}
                height={height}
                width={width}
                screenshotWidth={screenshotWidth}
                screenshotFormat={screenshotFormat}
                screenshotQuality={screenshotQuality}
            />
        );
    }
}

WebcamComponent.propTypes = {
    /**
     * The ID used to identify this component in Dash callbacks
     */
    id: PropTypes.string,

    /**
     * CSS class of video element
     */
    className: PropTypes.string,

    /**
     * enable/disable audio
     */
    audio: PropTypes.boolean,

    /**
     * height of video element
     */
    height: PropTypes.number,

    /**
     * width of video element
     */
    width: PropTypes.number,

    /**
     * width of screenshot
     */
    screenshotWidth: PropTypes.number,

    /**
     * format of screenshot
     */
    screenshotFormat: PropTypes.string,

    /**
     * quality of screenshot(0 to 1)
     */
    screenshotQuality: PropTypes.number,

    /**
     * update interval of screenshot in milliseconds
     */
    screenshotInterval: PropTypes.number,

    /**
     * getScreenshot() returns a base64 encoded string of the current webcam
     * image. screenshot is the prop updated at fixed interval from calling
     * getScreenshot(). It is currently capped at 25 fps.
     */
    screenshot: PropTypes.string,

    /**
     * Dash-assigned callback that should be called whenever any of the
     * properties change
     */
    setProps: PropTypes.func
};

WebcamComponent.defaultProps = {
    audio: true,
    className: '',
    height: 480,
    width: 640,
    screnshotFormat: 'image/webp',
    screenshotQuality: 0.92,
    screenshotInterval: 40
};

usage-advanced.py
"""
Note: audio and screenshotWidth can't be updated after it is set.
"""

import dash_webcam
import dash
from dash.dependencies import Input, Output, State
import dash_core_components as dcc
import dash_html_components as html

app = dash.Dash('')
server = app.server

app.scripts.config.serve_locally = True

app.layout = html.Div([
    html.Div(className='row', children=[
        html.P('Original Footage on the left, screenshot on the right'),

        dash_webcam.WebcamComponent(
            id='webcam',
            audio=True,
            screenshotFormat="image/jpeg",
            width=1280,
            height=720,
            screenshotWidth=700
        ),

        html.Img(id='output')
    ]),

    html.Div(className='row', style={'width': '50%'}, children=[
        html.P('Screenshot format:', style={'margin-top': '15px'}),
        dcc.RadioItems(
            id='radio-screenshot-format',
            options=[{'label': val.upper(), 'value': val} for val in [
                'webp',
                'jpeg',
                'png',
                'gif'
            ]],
            value='jpeg'
        ),

        html.P("Screenshot quality:", style={'margin-top': '30px'}),
        dcc.Slider(
            id='slider-screenshot-quality',
            min=0,
            max=1,
            step=0.1,
            value=0.8,
            updatemode='drag',
            marks={0: '0%', 1: '100%'}
        ),

        html.P("Update Interval for screenshots:",
               style={'margin-top': '30px'}),
        dcc.Slider(
            id='slider-interval-screenshot',
            min=40,
            max=1000,
            step=None,
            updatemode='drag',
            marks={i: str(i) for i in [40, 100, 200, 500, 1000]},
            value=100
        ),
    ])
])


@app.callback(Output('output', 'src'),
              [Input('webcam', 'screenshot')])
def show_screenshot_image(screenshot):
    return screenshot



@app.callback(Output('webcam', 'screenshotQuality'),
              [Input('slider-screenshot-quality', 'value')])
def update_screenshot_quality(value):
    return value


@app.callback(Output('webcam', 'screenshotInterval'),
              [Input('slider-interval-screenshot', 'value')])
def update_screenshot_interval(value):
    return value


@app.callback(Output('webcam', 'screenshotFormat'),
              [Input('radio-screenshot-format', 'value')])
def update_screenshot_format(value):
    return 'image/' + value


if __name__ == '__main__':
    app.run_server(debug=True)
